Add minimum working example

mattxwang opened the pull request:
This PR is a minimum working example (or MVP) for `stylelint-create`. Functionally, it has a simple workflow:

1. check if the user has the right version of node
2. check if a Stylelint config already exists
3. check if `package.json` exists
4. if all checks pass,
    1. install `stylelint` and `stylelint-config-standard` with the package manager's installer
    2. write a message for the user with next steps

In addition, I've:

- set up the typical "Stylelint stack" of ESLint, Prettier, Remark
- wrote a simple integration test for existing Stylelint config with Vitest
- run lint + test as actions

I've been manually testing with `node create-stylelint.mjs` in the project root, though I'm sure that there are better ways to do this.

I do have a few things I'd like some opinions/help on!

## Questions

High-level:

- should I be writing unit tests as well? or do we think integration tests are comprehensive in this case?
- I kept everything in one file since I felt that the logic wasn't too complex. Should I break things up (particularly, messages)?
- many of the ESLint norms we have aren't designed for CLI tools. Should I make adjustments locally (with comments), globally (disable rules in the config), or somewhere else?

Specific:

- should these integration tests be more comprehensive (ex examine the directory structure / last-modified times)? for now, it's mostly looking at error code + messages.
- in the tests, my approach to get the filepath for the executable and fixtures is very hacky. I struggled to find documentation on Vitest in the correct way to do this (it seems like most people are using it for Vite, or at least for web apps/server code rather than CLIs). Any thoughts?


## Items Left

Let me know if we think any of these are unnecessary, or if I am missing something.

- [ ] write more integration tests:
    - [ ] happy path
    - [x] `package.json` doesn't exist
    - [ ] failure during installation (I think I'll have to mock `execa` for this)
- [x] add TypeScript
- [x] test on macOS, Windows (I was planning on doing this through CI)
- [ ] better documentation

I'd love to keep working on this and add other features soon!

---

Ref: https://github.com/stylelint/stylelint/issues/4269.

ybiquitous commented on create-stylelint.mjs:
```diff
@@ -0,0 +1,17 @@
+const requiredMajorVersion = parseInt(currentVersion.split('.')[0], 10);
+const minimumMajorVersion = 14;
+
+if (requiredMajorVersion < minimumMajorVersion) {
```
[suggest] Checking the running version sounds useful. 👍🏼 
Additionally, by using the `engines.node` field in `package.json`, I believe it is possible to make this check more accurate, though code and dependency (`semver`) increase a bit. For example:

```js
import { fileURLToPath } from 'node:url';
import fs from 'node:fs';
import path from 'node:path';
import semver from 'semver';

const dirname = path.dirname(fileURLToPath(import.meta.url));
const pkg = JSON.parse(fs.readFileSync(path.join(dirname, 'package.json').toString()));

if (!semver.satisfies(currentVersion, pkg.engines.node)) {
	console.error(`Node.js v${currentVersion} is out of date and unsupported!`);
	console.error(`Please use a Node.js version within "${pkg.engines.node}".`);
	process.exit(1);
}
```

ybiquitous commented on package.json:
```diff
@@ -0,0 +1,67 @@
+  },
+  "scripts": {
+    "format": "prettier . --write --cache",
+    "jest": "jest",
```
[trivial] Maybe unnecessary.
```suggestion
```

ybiquitous commented on package.json:
```diff
@@ -0,0 +1,67 @@
+  },
+  "license": "MIT",
+  "author": "stylelint",
+  "main": "create-stylelint.mjs",
```
[trivial] Perhaps `main` is unnecessary because this package is pure ESM.
```suggestion
  "main": "create-stylelint.mjs",
```

mattxwang replied:
Removing this causes triggers the [`require-main`](https://npmpackagejsonlint.org/docs/rules/required-node/require-main) rule in `npm-package-json-lint`. Should I keep the `main` entry, or disable the rule?

ybiquitous replied:
OK. Let's keep it as-is.

ntwb replied:
If need be we can ignore this rule for this package by adding this to the `npmpackagejsonlint` config object:

```
"overrides": [
    {
      "rules": {
        "require-main": "off"
      }
    }
  ]
```

ybiquitous commented on package.json:
```diff
@@ -0,0 +1,67 @@
+    "@stylelint/remark-preset": "^4.0.0",
+    "eslint": "^8.28.0",
+    "eslint-config-stylelint": "^17.1.0",
+    "jest": "^28.1.3",
+    "jest-preset-stylelint": "^5.0.4",
```
[suggest] Jest is unnecessary.

```suggestion
```

mattxwang replied:
Ah, so this is what I noted above in the PR description; if I remove this, e.g. [see this commit's action run](https://github.com/stylelint/stylelint-create/actions/runs/4038050780/jobs/6941775169), then `eslint-plugin-jest` fails because `jest` doesn't exist. 

I assume the path forward is to disable the plugin, but "unextending" seems trickier than I thought. Any suggestions?

ntwb replied:
One option is to set the Jest version in `package.json`:

```
"settings": {
    "jest": {
      "version": 26
    }
 }
```

It also appears this issue only affects the ESLInt `jest/no-deprecated-functions` rule

Discussion of this can be seen here https://github.com/jest-community/eslint-plugin-jest/issues/686

An additional suggestion is to use `latest` as the version ,if `latest` works this would be far more maintainable and we wouldn't have to update this ourselves with each new Jest major release

https://github.com/jest-community/eslint-plugin-jest/issues/686#issuecomment-739151588
> A workaround for this issue is to specify the version via `latest `as most ESLint versions are back-wards compatible in the `.eslintrc`.

ybiquitous replied:
Oh, our ESLint config should be updated to exclude the configs for Jest:

https://github.com/stylelint/eslint-config-stylelint/blob/43802c04eef4134680a58205271c8111a05f0434/.eslintrc.js#L16-L17

The `jest/no-deprecated-functions` rule may require that the Jest package should be installed, so could you disable only the rule to suppress the error like this?

```diff
--- a/package.json
+++ b/package.json
@@ -46,6 +46,9 @@
     "extends": [
       "stylelint"
     ],
+    "rules": {
+      "jest/no-deprecated-functions": "off"
+    },
     "root": true
   },
```

ybiquitous replied:
Sorry, I missed @ntwb's comment before I posted. It also seems reasonable to set `settings.jest.version` as a workaround.

I've opened an issue related to this problem: https://github.com/stylelint/eslint-config-stylelint/issues/205

mattxwang replied:
Ended up setting to `latest` for now; happy to resolve later once we have a more permanent solution!

ybiquitous commented on src/index.mjs:
```diff
@@ -0,0 +1,92 @@
+const DEFAULT_CONFIG_FILE = '.stylelintrc.json';
+
+const STYLELINT_CONFIG_PATHS = new Set([
+	'.stylelintrc',
```
[suggest] Adding also `.cjs` seems useful:

```suggestion
	'.stylelintrc',
	'.stylelintrc.cjs',
```

ntwb replied:
Potentially additional files and paths based on cosmiconfig: 

https://github.com/davidtheclark/cosmiconfig/#cosmiconfig

- "any of the above two inside a `.config` subdirectory"
- "a `.config.js` or `.config.cjs` CommonJS module"

ybiquitous commented on src/index.mjs:
```diff
@@ -0,0 +1,92 @@
+
+	if (existingConfigs.length > 0) {
+		console.error(
+			chalk.red(`The ${existingConfigs} config(s) already exist. Remove them and then try again.`),
```
[suggest] A stringified array may be a bit hart to read because each element is joined by a comma (,) without whitespace. What about using `.join()`?

```suggestion
			chalk.red(`The ${existingConfigs.join(', ')} config(s) already exist. Remove them and then try again.`),
```

ybiquitous commented on src/index.mjs:
```diff
@@ -0,0 +1,92 @@
+
+	if (!directoryHasPackageJson(cwd)) {
+		console.error(
+			chalk.red(`The package.json was not found. Run ${pkgManager} init and then try again.`),
```
[suggest] Adding quotes for a command seems a bit readable:
```suggestion
			chalk.red(`The package.json was not found. Run "${pkgManager} init" and then try again.`),
```

ybiquitous commented on test/integration.test.js:
```diff
@@ -0,0 +1,30 @@
+}
+
+function setup(pathToTest, projectRoot, args = []) {
+	return execaSync('node', [`${projectRoot}create-stylelint.mjs`, ...args], {
```
[suggest] Possible to simplify the code by adding a execute permission to the file:

```console
$ chmod +x create-stylelint.mjs
```

```suggestion
	return execaSync(`${projectRoot}create-stylelint.mjs`, args, {
```

mattxwang replied:
It seems like this is necessary for this to work on Windows (I tried the suggestion, and it failed on CI). I'm not familiar with windows development, so perhaps I'm missing something?

ybiquitous replied:
Ah, sounds good. Let's keep it. Thanks.

ybiquitous commented on test/integration.test.js:
```diff
@@ -0,0 +1,30 @@
+	it('should not proceed if a Stylelint config already exists in the directory', (context) => {
+		const projectRoot = getProjectRoot(context);
+
+		try {
+			setup(inputs.stylelintConfigExists, projectRoot);
+		} catch (error) {
+			expect(error.failed).toBe(true);
+			expect(error.message).toMatch('config(s) already exist');
+		}
```
[suggest] Possible to avoid a test bug (that a `try` block unexpectedly passes) by using `.toThrowError`. For example:

```suggestion
		expect(() => setup(inputs.stylelintConfigExists, projectRoot)).toThrowError(/config\(s\) already exist/)
```

ybiquitous reviewed (changes requested):
Thanks for creating this MVP. It seems great as a basis!

ntwb commented on LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
This file can be removed and it will be inherited from the shared repo at https://github.com/stylelint/.github/blob/main/LICENSE

ntwb commented on README.md:
```diff
@@ -1 +1,23 @@
+
+Commits to `main` will be tested by [GitHub Actions](https://github.com/stylelint/create-stylelint/actions)
+
+## License
+
+This repository is licensed under the MIT License.
+
+The development of this project is heavily inspired by [create-astro](https://github.com/withastro/astro/blob/main/packages/create-astro); thank you to the Astro team!
```
As per my above comment on removing the LICENSE file, this is also probably not required as the license will be displayed on the repo as it is for the main Stylelint repo: https://github.com/stylelint/stylelint

![image](https://user-images.githubusercontent.com/1016458/214996420-afaf5b06-6142-43b8-b01f-740189fe4a42.png)

ntwb commented on package.json:
```diff
@@ -0,0 +1,67 @@
+  },
+  "devDependencies": {
+    "@stylelint/prettier-config": "^2.0.0",
+    "@stylelint/remark-preset": "^4.0.0",
```
Please also add https://github.com/stylelint/npm-package-json-lint-config
- `npm install @stylelint/npm-package-json-lint-config --save-dev`

And the accompaning config object here in this file:
- `"npmpackagejsonlint": "@stylelint/npm-package-json-lint-config"`


And the script to run the checks:
- `"lint:packagejson": "npmPkgJsonLint .",`

ntwb reviewed (commented):
Looking great, added some comments...

mattxwang commented on package.json:
```diff
@@ -0,0 +1,67 @@
+    "vitest": "^0.28.1"
+  },
+  "engines": {
+    "node": "^14.13.1 || >=16.0.0"
```
I believe so, I'll add it to the matrix!

jeddy3 replied:
> [ask] Is it necessary to support Node.js v14?

We should align with whatever we do for `stylelint@15.0.0`. We're currently dropping support for node@12, but node@14 is EOL in a couple of months' time. Maybe we should drop that too as part of `stylelint@15.0.0`?

ybiquitous replied:
> Maybe we should drop that too as part of `stylelint@15.0.0`?

I think it is a bit premature, and Stylelint 16 is good timing to drop Node.js 14.

mattxwang commented on src/index.mjs:
```diff
@@ -0,0 +1,92 @@
+async function installPackages(cwd, pkgManager) {
+	const installExec = execa(
+		pkgManager,
+		[`${getInstallCommand(pkgManager)}`, '-D', 'stylelint', 'stylelint-config-standard'],
```
I believe so - see the docs for [yarn add](https://yarnpkg.com/cli/add#options-D%2C-dev) and [pnpm add](https://pnpm.io/cli/add#--save-dev--d).

ybiquitous replied:
I see. Thanks.

jeddy3 commented on .gitignore:
```diff
@@ -0,0 +1,106 @@
+# Logs
```
Will the following suffice?:

```
node_modules
*.log
```

mattxwang replied:
Yup; I used the default provided by GitHub, happy to slim it down. I'll merge your and @Mouvedia's suggestion.

jeddy3 commented on README.md:
```diff
@@ -1 +1,21 @@
\ No newline at end of file
+# stylelint-create
+
+The official (in-development) [`npm init`](https://docs.npmjs.com/cli/v9/commands/npm-init) script for [Stylelint](https://github.com/stylelint/stylelint).
+
+Tracking issue: https://github.com/stylelint/stylelint/issues/4269
+
```
```suggestion
The [`npm init`](https://docs.npmjs.com/cli/v9/commands/npm-init) script for [Stylelint](https://github.com/stylelint/stylelint).

```

Getting ready to release the first version.

jeddy3 commented on README.md:
```diff
@@ -1 +1,21 @@
+
+Tracking issue: https://github.com/stylelint/stylelint/issues/4269
+
+## Use the tool
+
+Coming soon!
```
```suggestion
## Usage

Run `npm init stylelint`
```

jeddy3 commented on README.md:
```diff
@@ -1 +1,21 @@
+
+Coming soon!
+
+## Getting started
+
+- `npm install`
+- to test locally, run `node create-stylelint.mjs`
+- to test with [Vitest](https://vitest.dev/), `npm test`
+
+Commits to `main` will be tested by [GitHub Actions](https://github.com/stylelint/create-stylelint/actions)
+
```
```suggestion
## Contributing

- `npm install`
- to test locally, run `node create-stylelint.mjs`
- to test with [Vitest](https://vitest.dev/), `npm test`

```

jeddy3 commented on README.md:
```diff
@@ -1 +1,21 @@
+
+## Acknowledgements
+
+The development of this project is heavily inspired by [create-astro](https://github.com/withastro/astro/blob/main/packages/create-astro); thank you to the Astro team!
```
```suggestion
Inspired by [create-astro](https://github.com/withastro/astro/blob/main/packages/create-astro).
```

jeddy3 commented on create-stylelint.mjs:
```diff
@@ -0,0 +1,24 @@
+const pkg = JSON.parse(fs.readFileSync(path.join(dirname, 'package.json').toString()));
+
+if (!semver.satisfies(currentVersion, pkg.engines.node)) {
+	console.error(`Node.js v${currentVersion} is out of date and unsupported!`);
```
```suggestion
	console.error(`Unsupported Node.js version (v${currentVersion})`);
```

jeddy3 commented on create-stylelint.mjs:
```diff
@@ -0,0 +1,24 @@
+
+if (!semver.satisfies(currentVersion, pkg.engines.node)) {
+	console.error(`Node.js v${currentVersion} is out of date and unsupported!`);
+	console.error(`Please use a Node.js version within "${pkg.engines.node}".`);
```
```suggestion
	console.error(`Install a Node.js version within "${pkg.engines.node}" and then try again.`);
```

To be consistent with other error messages.

jeddy3 commented on package.json:
```diff
@@ -0,0 +1,81 @@
+  "version": "0.0.1",
+  "description": "Initialization wizard for Stylelint, a linter that helps you avoid errors and enforce conventions in your styles.",
+  "keywords": [
+    "css-in-js",
+    "css",
+    "less",
+    "lint",
+    "linter",
+    "markdown",
+    "sass",
+    "scss",
+    "stylelint",
+    "sugarss"
```
```suggestion
    "stylelint",
    "npm init"
```

Mouvedia replied:
```suggestion
    "cli",
    "create",
    "init",
    "generator",
    "scaffolding",
    "template"
```

jeddy3 commented on package.json:
```diff
@@ -0,0 +1,81 @@
+    "stylelint",
+    "sugarss"
+  ],
+  "homepage": "https://github.com/stylelint/stylelint-create#readme",
+  "bugs": {
+    "url": "https://github.com/stylelint/stylelint-create/issues"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/stylelint/stylelint-create.git"
+  },
```
```suggestion
  "repository": "stylelint/create-stylelint",
```

Fields are inferred.

jeddy3 commented on src/index.mjs:
```diff
@@ -0,0 +1,96 @@
+]);
+
+const NEXT_STEPS_STRING = `
+You can now lint your CSS files using: npx stylelint "**/*.css"
+
```
```suggestion
You can now lint your CSS files using:
npx stylelint "**/*.css"

```

jeddy3 commented on src/index.mjs:
```diff
@@ -0,0 +1,96 @@
+			installSpinner.text = `${installingPackagesMsg}\n${chalk.bold(`[${pkgManager}]`)} ${data}`;
+		});
+		installExec.on('error', (error) => {
+			console.error(chalk.red(`Error installing packages: ${error}`));
```
```suggestion
			console.error(chalk.red(`Failed to install packages: ${error}`));
```

jeddy3 commented on src/index.mjs:
```diff
@@ -0,0 +1,96 @@
+		});
+		installExec.on('close', () => resolve());
+	});
+	installSpinner.text = chalk.green('Packages installed.');
```
```suggestion
	installSpinner.text = chalk.green('Installed packages.');
```

jeddy3 reviewed (changes requested):
Looking good!

I've made some suggestions.

I don't think we're far off having this ready. It'd be nice to roll it into the v15 launch.

mattxwang commented:
Thanks everyone for chiming in - will get to addressing reviews soon! Just to be transparent, won't be able to work on this likely until Friday or the weekend (catching up after travelling for interviews)!

Mouvedia commented on package.json:
```diff
@@ -0,0 +1,81 @@
+    "lint:formatting": "prettier . --check --cache",
+    "lint:js": "eslint . --cache --max-warnings=0 --ext .js,.mjs",
+    "lint:md": "remark . --quiet --frail",
+    "lint:packagejson": "npmPkgJsonLint .",
```
I didn't know about that one.
Thanks for the find!

Mouvedia commented on vitest.config.ts:
```diff
@@ -0,0 +1,7 @@
+
+export default defineConfig({
+	test: {
+		include: ['**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx,json}'],
```
```suggestion
		include: ['**/*.test.{js,mjs,cjs}'],
```

Mouvedia commented on .gitignore:
```diff
@@ -0,0 +1,106 @@
+# Logs
```
```suggestion
*~
*[._]sw[g-p]
.eslintcache

node_modules/
.idea/
.vscode/
```

I agree with @jeddy3.
KISS

Mouvedia commented on package.json:
```diff
@@ -0,0 +1,81 @@
+    "create-stylelint": "./create-stylelint.mjs"
+  },
+  "scripts": {
+    "format": "prettier . --write --cache",
```
question

Would `--ignore-path .gitignore` be necessary?

mattxwang replied:
Is your concern `node_modules`? If so, that's handled [by default](https://prettier.io/docs/en/ignore.html):

> By default prettier ignores files in version control systems directories (".git", ".svn" and ".hg") and node_modules (if [--with-node-modules CLI option](https://prettier.io/docs/en/cli.html#--with-node-modules) not specified)

ybiquitous commented on package.json:
```diff
@@ -0,0 +1,72 @@
+    "stylelint": "^14.16.1",
+    "stylelint-config-standard": "^29.0.0",
+    "vitest": "^0.28.1"
+  },
```
[ask] `tsconfig.json` is committed, but `typescript` is not installed into `devDependencies`. Do you think it is intentional? If so, what is the reason?

Mouvedia replied:
That's probably because has https://github.com/withastro/astro/tree/main/packages/create-astro a `--typescript` flag.
see https://github.com/stylelint/stylelint-create/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R17

So Id say convert it to a simple `jsconfig.json`.
e.g.
```json
{
  "compilerOptions": {
    "checkJs": true
  }
}
```

mattxwang replied:
Ah, I was relying on typescript implicitly installed with some of the other deps, but didn't explicitly put it as a devDep in `package.json`.

Originally, the idea was to use TS for the `src/` files - see the annotations in `index.mjs`. I had thought that this setup is what most closely matches Stylelint (ex `.js` extensions, but inline TS). Any suggestions on what I should do? I can:

- formally add TS, and set up a bit more infrastructure to make it obvious/easy to use; unsure if i need to, but I can also add a compilation/build step
- remove the `tsconfig.json` and use a `jsconfig.json` as suggested
- something else!

ybiquitous replied:
I understand the reason, thanks. I have no problem keeping `tsconfig.json`, while it seems good to me to install `typescript` as dev dep and add `scripts.lint:types` as `stylelint` does. Though, the latter is nice to have.

I disagree with building source files (`*.ts`) because I want to be able to run them as they are.

mattxwang replied:
Sounds good - I'll do that!

ybiquitous reviewed (commented):
I've left one question, but SGTM. Thanks!

jeddy3 commented on src/index.mjs:
```diff
@@ -0,0 +1,117 @@
+	console.log(picocolors.green(`Created ${DEFAULT_CONFIG_FILE}.`));
+
+	await installPackages(cwd, pkgManager);
+
+	console.log(picocolors.green('Stylelint has been fully configured.'));
```
```suggestion
```

Let's remove this redundant message.

jeddy3 reviewed (changes requested):
Looking great! 

I've added a minor suggestion.

Mouvedia commented on src/index.mjs:
```diff
@@ -0,0 +1,115 @@
+/* eslint-disable no-process-exit */
+/* eslint no-console: 'off' */
```
off topic

maybe we could add `"assert", "debug", "info", "exception"` to https://github.com/stylelint/eslint-config-stylelint/blob/4475504c22be754d30e529edfb7ad5db5765adf9/.eslintrc.js#LL37C19-L37C19

ybiquitous replied:
Because we should detect debug messages or assertions with `console`, I think keeping it as-is is reasonable.

ybiquitous reviewed (approved):
Great work! LGTM 👍🏼

jeddy3 reviewed (approved):
Awesome stuff. I think we can merge this MVP and then pick up any improvements in new tickets and PRs.
